relay-event-schema/src/protocol/span/compat.rs

```diff
@@ -0,0 +1,240 @@
+use std::str::FromStr;
+
+use relay_protocol::{Annotated, Empty, FromValue, IntoValue, Object, Value};
+
+use crate::protocol::{EventId, SpanV2, Timestamp};
+
+/// Temporary type that amends a SpansV2 span with fields needed by the sentry span consumer.
+/// This can be removed once the consumer has benn updated to use the new schema.
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue)]
+#[allow(dead_code)]
+pub struct CompatSpan {
+    #[metastructure(flatten)]
+    pub span_v2: SpanV2,
+
+    pub data: Annotated<Object<Value>>,
+    pub description: Annotated<String>,
+    pub duration_ms: Annotated<u64>,
+    pub end_timestamp_precise: Annotated<Timestamp>,
+    pub profile_id: Annotated<EventId>,
+    pub segment_id: Annotated<String>,
+    pub start_timestamp_ms: Annotated<u64>, // TODO: remove from kafka schema, no longer used in consumer
+    pub start_timestamp_precise: Annotated<Timestamp>,
+}
+
+impl TryFrom<SpanV2> for CompatSpan {
+    type Error = uuid::Error;
+
+    fn try_from(span_v2: SpanV2) -> Result<Self, uuid::Error> {
+        let mut compat_span = Self::default();
+
+        if let Some(start_timestamp) = span_v2.start_timestamp.value() {
+            let dt = start_timestamp.0;
+            compat_span.start_timestamp_precise = (*start_timestamp).into();
+            compat_span.start_timestamp_ms = (dt.timestamp_millis() as u64).into();
+        }
+
+        if let Some(end_timestamp) = span_v2.end_timestamp.value() {
+            compat_span.end_timestamp_precise = (*end_timestamp).into();
+        }
+
+        if let (Some(start_timestamp), Some(end_timestamp)) = (
+            span_v2.start_timestamp.value(),
+            span_v2.end_timestamp.value(),
+        ) {
+            let delta = (*end_timestamp - *start_timestamp).num_milliseconds();
+            compat_span.duration_ms = u64::try_from(delta).unwrap_or(0).into();
+        }
+
+        if let Some(attributes) = span_v2.attributes.value() {
+            for (key, value) in attributes.iter() {
+                if let Some(attribute) = value.value()
+                    && let Some(value) = attribute.value.value.value()
+                {
```

Comment:
This could also be done with `filter_map`.
```suggestion
            for (key, value) in attributes
                .iter()
                .filter_map(|(k, v)| (k, v.value()?.value.value.value()))
```
Or we could introduce an iterator on `Attributes` that returns `(&str, &Value)` (i.e. exactly what is used here). Wouldn't know what to call it, though.

Reply:
I used `filter_map` for now, let's see if the use case comes up more often.